File: epi_judge_java/epi/QueueWithMax.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiUserType;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;

import java.util.ArrayDeque;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Queue;
import java.util.TreeMap;

public class QueueWithMax {
    private final Queue<Integer> q = new ArrayDeque<>();
    private final TreeMap<Integer, Integer> numToCountMap = new TreeMap<>((a, b) -> b - a);

    public void enqueue(Integer x) {
        q.offer(x);
        numToCountMap.merge(x, 1, Integer::sum);
    }

    public Integer dequeue() {
        Integer polled = q.poll();

        Integer count = numToCountMap.get(polled);
        if (count == 1) {
            numToCountMap.remove(polled);
        } else {
            numToCountMap.put(polled, count - 1);
        }

        return polled;
    }

    public Integer max() {
        return numToCountMap.firstKey();
    }

    @EpiUserType(ctorParams = {String.class, int.class})
    public static class QueueOp {
        public String op;
        public int arg;

        public QueueOp(String op, int arg) {
            this.op = op;
            this.arg = arg;
        }
    }

    @EpiTest(testDataFile = "queue_with_max.tsv")
    public static void queueTester(List<QueueOp> ops) throws TestFailure {
        try {
            QueueWithMax q = new QueueWithMax();

            for (QueueOp op : ops) {
                switch (op.op) {
                    case "QueueWithMax":
                        q = new QueueWithMax();
                        break;
                    case "enqueue":
                        q.enqueue(op.arg);
                        break;
                    case "dequeue":
                        int result = q.dequeue();
                        if (result != op.arg) {
                            throw new TestFailure("Dequeue: expected " +
                                    String.valueOf(op.arg) + ", got " +
                                    String.valueOf(result));
                        }
                        break;
                    case "max":
                        int s = q.max();
                        if (s != op.arg) {
                            throw new TestFailure("Max: expected " + String.valueOf(op.arg) +
                                    ", got " + String.valueOf(s));
                        }
                        break;
                }
            }
        } catch (NoSuchElementException e) {
            throw new TestFailure("Unexpected NoSuchElement exception");
        }
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "QueueWithMax.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/queue_with_max.cc
#include <stdexcept>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
#include "test_framework/test_failure.h"

using namespace std;

class QueueWithMax {
public:
    void Enqueue(int x) {
        q.push(x);

        while (!maxDeque.empty() && maxDeque.back() < x) {
            maxDeque.pop_back();
        }

        maxDeque.emplace_back(x);
    }

    int Dequeue() {
        int result = q.front();
        q.pop();

        if (result == maxDeque.front()) {
            maxDeque.pop_front();
        }

        return result;
    }

    int Max() const {
        return maxDeque.front();
    }

private:
    queue<int> q;
    deque<int> maxDeque;
};

struct QueueOp {
    enum class Operation {
        kConstruct, kDequeue, kEnqueue, kMax
    } op;
    int argument;

    QueueOp(const std::string &op_string, int arg) : argument(arg) {
        if (op_string == "QueueWithMax") {
            op = Operation::kConstruct;
        } else if (op_string == "dequeue") {
            op = Operation::kDequeue;
        } else if (op_string == "enqueue") {
            op = Operation::kEnqueue;
        } else if (op_string == "max") {
            op = Operation::kMax;
        } else {
            throw std::runtime_error("Unsupported queue operation: " + op_string);
        }
    }
};

namespace test_framework {
    template<>
    struct SerializationTrait<QueueOp> : UserSerTrait<QueueOp, std::string, int> {
    };
}  // namespace test_framework

void QueueTester(const std::vector<QueueOp> &ops) {
    try {
        QueueWithMax q;
        for (auto &x: ops) {
            switch (x.op) {
                case QueueOp::Operation::kConstruct:
                    break;
                case QueueOp::Operation::kDequeue: {
                    int result = q.Dequeue();
                    if (result != x.argument) {
                        throw TestFailure("Dequeue: expected " +
                                          std::to_string(x.argument) + ", got " +
                                          std::to_string(result));
                    }
                }
                    break;
                case QueueOp::Operation::kEnqueue:
                    q.Enqueue(x.argument);
                    break;
                case QueueOp::Operation::kMax: {
                    int s = q.Max();
                    if (s != x.argument) {
                        throw TestFailure("Max: expected " + std::to_string(x.argument) +
                                          ", got " + std::to_string(s));
                    }
                }
                    break;
            }
        }
    } catch (const length_error &) {
        throw TestFailure("Unexpected length_error exception");
    }
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"ops"};
    return GenericTestMain(args, "queue_with_max.cc", "queue_with_max.tsv",
                           &QueueTester, DefaultComparator{}, param_names);
}
